### PR TITLE
LauncherActivity allows async tasks to run before launching

### PR DESCRIPTION
**Pull request by andreban**
- LauncherActivity allows async tasks to run before launching
- Adds a FirebaseAnalytics demo

### CODE REVIEW

File: demos/twa-firebase-analytics/build.gradle [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            minifyEnabled false
+        }
+    }
+}
+
+dependencies {
+    implementation fileTree(dir: "libs", include: ["*.jar"])
+    implementation project(path: ':androidbrowserhelper')
+    implementation 'androidx.browser:browser:1.3.0'
+    // Firebase : firebase bom for automatic choice version for each imported library
```
Comment (PEConn) [inline demos/twa-firebase-analytics/build.gradle:53]: What's a bom?

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java [hunk 2]
```diff
@@ -134,6 +142,16 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
             return;
         }
 
+        if (mLaunchImmediately) {
+            this.launchTwa();
```
Comment (PEConn) [inline androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java:146]: nit: You don't need `this` here. (Too much TS?)

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java [hunk 3]
```diff
@@ -118,6 +118,14 @@
     @Nullable
     private TwaLauncher mTwaLauncher;
 
+    /**
+     * Signals if `LauncherActivity` should automatically launch the Trusted Web Activity on
+     * `onCreate()`. Set to `false` when a subclass needs to perform an asynchronous task before
+     * launching the Trusted Web Activity. The subclass will then be responsible for calling
+     * `launch()` itself once the asynchronous task is finished.
+     */
+    protected boolean mLaunchImmediately = true;
```
Comment (PEConn) [inline androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java:127]: I don't like protected variables, usually we can get by with mutable private variables and immutable, non-private variables. (I can probably come up with a more concrete reason I don't like them if you want one.)

I can think of two ways that I think would be better:
1. Create a `boolean shouldLaunchImmediately()` method that can be overridden.
2. Create some protected method like `void onPostCreate()` that gets called at the end of `onCreate` and which by default calls `launchTwa()`. The user can then override this to be empty and call `launchTwa` themselves. (Admittedly the naming here is a bit ugly.)

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import com.google.android.gms.tasks.Task;
+import com.google.androidbrowserhelper.trusted.LauncherActivity;
+import com.google.firebase.analytics.FirebaseAnalytics;
+
+public class FirebaseAnalyticsLauncherActivity extends LauncherActivity {
+    private String mAppInstanceId;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.mLaunchImmediately = false;
```
Comment (PEConn) [inline demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java:33]: You don't need `super` on this line (you do on the next line though).

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.androidbrowserhelper.trusted.LauncherActivity;
+import com.google.firebase.analytics.FirebaseAnalytics;
+
+public class FirebaseAnalyticsLauncherActivity extends LauncherActivity {
```
Comment (PEConn) [inline demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java:28]: nit: javadoc on public classes/methods (unless they're overrides).

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    private String mAppInstanceId;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.mLaunchImmediately = false;
+        super.onCreate(savedInstanceState);
+
+        FirebaseAnalytics mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);
```
Comment (PEConn) [inline demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java:36]: nit: mVariableName is for member variables, just "firebaseAnalytics" please.

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+        // Start the asynchronous task to get the Firebase application instance id.
+        mFirebaseAnalytics.getAppInstanceId().addOnCompleteListener(new OnCompleteListener<String>() {
```
Comment (PEConn) [inline demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java:39]: You can use a lambda here:

```
...addOnCompleteListener(task -> {
  // Do stuff with task.
});
```

(you may need to enable Java 8 in gradle, but we use lambdas elsewhere)

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            @Override
+            public void onComplete(@NonNull Task<String> task) {
+                // Once the task is complete, save the instance id so it can be used by
+                // getLaunchingUrl().
+                mAppInstanceId = task.getResult();
+                FirebaseAnalyticsLauncherActivity.this.launchTwa();
```
Comment (PEConn) [inline demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java:45]: Once you're in a lambda, you won't need `FirebaseAnalyticsLauncherActivity.this`, `launchTwa()` should just work.

**Review by PEConn — COMMENTED** [review]
This is good. One optimization for a follow up PR would be to kick off the TWA service connection in onCreate. That way we do our two async tasks in parallel (connecting to the service and connecting to Firebase) instead of one after the other.

File: demos/twa-firebase-analytics/build.gradle (hunk 1 above)
Comment (andreban) [reply demos/twa-firebase-analytics/build.gradle]: "Bill of Materials": https://firebase.google.com/docs/android/learn-more#bom. Updated the comment.

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java (hunk 2 above)
Comment (andreban) [reply androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: Yes... too much TS, but TBH, I already had the habit of using `this` in Java even when not necessary.  Removed.

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java (hunk 3 above)
Comment (andreban) [reply androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: Moved to a `shouldLaunchImmediately()` method.

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java (hunk 4 above)
Comment (andreban) [reply demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java]: Removed this line.

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java (hunk 5 above)
Comment (andreban) [reply demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java:32]: Added

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java (hunk 6 above)
Comment (andreban) [reply demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java]: Renamed the variable.

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java (hunk 7 above)
Comment (andreban) [reply demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java]: Done.

File: demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java (hunk 8 above)
Comment (andreban) [reply demos/twa-firebase-analytics/src/main/java/com/google/androidbrowserhelper/demos/twa_firebase_analytics/FirebaseAnalyticsLauncherActivity.java]: Yep, using only launchTwa () now.

**Review by jackiebrewster669 — APPROVED** [review]
/data/user/0/com.aide.ui/files/Lessons/Android Hello World/res/menu/rootscces.xml

**Comment by Chainarong24NutDogg** [comment]
เมื่อ 4 ม.ค. 2021 13:39 น. jackiebrewster669 <notifications@github.com> เขียนว่า
@jackiebrewster669 approved this pull request.

/data/user/0/com.aide.ui/files/Lessons/Android Hello World/res/menu/rootscces.xml

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java [hunk 9]
```diff
@@ -134,6 +134,26 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
             return;
         }
 
+        if (shouldLaunchImmediately()) {
+            launchTwa();
+        }
+    }
+
+    /**
+     * Signals if `LauncherActivity` should automatically launch the Trusted Web Activity on
```
Comment (PEConn) [inline androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java:143]: Unfortunately markdown doesn't work in javadoc comments - we've got to use a whole different annotation system.

Basically, if you want something to be code, do `{@code something}`. If you want to reference a class, do `{@link LauncherActivity}` and if you want to reference a method, do `{@link #shouldLaunchImmediately}`.

**Review by PEConn — APPROVED** [review]
LGTM with 1 doc nit.

File: androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java (hunk 9 above)
Comment (andreban) [reply androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/LauncherActivity.java]: 🤦  fixed

**Review by jackiebrewster669 — COMMENTED** [review]
``andreban:firebase-analytics[andreban:firebase-analytics](andreban:firebase-analytics)

**Pull request merged**